internal/datasource/nginx/log_tailer.go

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) F5, Inc.
+//
+// This source code is licensed under the Apache License, Version 2.0 license found in the
+// LICENSE file in the root directory of this source tree.
+
+package nginx
+
+import (
+	"context"
+	"io"
+	"log/slog"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/nxadm/tail"
+	"github.com/trivago/grok"
+)
+
+const numberOfTimesToSplitLogLine = 2
+
+var tailConfig = tail.Config{
+	Follow:    true,
+	ReOpen:    true,
+	MustExist: true,
+	Poll:      true,
+	Location: &tail.SeekInfo{
+		Whence: io.SeekEnd,
+	},
+}
+
+// NginxAccessItem represents the decoded access log data
+type NginxAccessItem struct {
+	BodyBytesSent          string `mapstructure:"body_bytes_sent"`
+	Status                 string `mapstructure:"status"`
+	RemoteAddress          string `mapstructure:"remote_addr"`
+	HTTPUserAgent          string `mapstructure:"http_user_agent"`
+	Request                string `mapstructure:"request"`
+	BytesSent              string `mapstructure:"bytes_sent"`
+	RequestLength          string `mapstructure:"request_length"`
+	RequestTime            string `mapstructure:"request_time"`
+	GzipRatio              string `mapstructure:"gzip_ratio"`
+	ServerProtocol         string `mapstructure:"server_protocol"`
+	UpstreamConnectTime    string `mapstructure:"upstream_connect_time"`
+	UpstreamHeaderTime     string `mapstructure:"upstream_header_time"`
+	UpstreamResponseTime   string `mapstructure:"upstream_response_time"`
+	UpstreamResponseLength string `mapstructure:"upstream_response_length"`
+	UpstreamStatus         string `mapstructure:"upstream_status"`
+	UpstreamCacheStatus    string `mapstructure:"upstream_cache_status"`
+}
+
+func NewNginxAccessItem(v map[string]string) (*NginxAccessItem, error) {
+	res := &NginxAccessItem{}
+	if err := mapstructure.Decode(v, res); err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+type Tailer struct {
+	handle *tail.Tail
+}
+
+type PatternTailer struct {
+	handle *tail.Tail
+	gc     *grok.CompiledGrok
+}
+
+type LTSVTailer struct {
+	handle *tail.Tail
+}
```

Comment:
Should we group all the `type`s in the same block? i.e. move them with `NginxAccessItem`

Reply:
yeah can do